File: src/c/appendix/rain_tier.c
// src/c/appendix/rain_tier.c
#include "rain_tier.h"
#include "palette.h"

// Inclusive upper bounds in wire tenths for tiers 1..4. Tier 5 catches the rest.
static const int RAIN_TIER_MAX_TENTHS[RAIN_TIER_COUNT - 1] = { 1, 5, 20, 100 };

// Top of each slab as a cumulative percentage of plot_h, indexed by tier
// (0 = axis, RAIN_TIER_TOP_PCT[k] = top of slab k). Tier 1 occupies the
// bottom 14%, tier 2 the next 20%, and tiers 3..5 22% each.
static const int RAIN_TIER_TOP_PCT_ARR[RAIN_TIER_COUNT + 1] = { 0, 14, 34, 56, 78, 100 };

int rain_tier_of_tenths(int tenths) {
    if (tenths <= 0) {
        return 0;
    }
    for (int i = 0; i < RAIN_TIER_COUNT - 1; ++i) {
        if (tenths <= RAIN_TIER_MAX_TENTHS[i]) {
            return i + 1;
        }
    }
    return RAIN_TIER_COUNT;
}

static int rain_tier_fill_q8(int tenths, int tier) {
    int low, high;
    switch (tier) {
        case 1: return 256;
        case 2: low = 2;   high = 5;   break;
        case 3: low = 6;   high = 20;  break;
        case 4: low = 21;  high = 100; break;
        case 5: low = 101; high = 255; break;
        default: return 256;
    }
    if (tenths >= high) { return 256; }
    if (tenths <= low)  { return 0; }
    return ((tenths - low) * 256) / (high - low);
}

int rain_tier_pixel_height(int tier, int bar_plot_h) {
    if (tier <= 0 || bar_plot_h <= 0) {
        return 0;
    }
    int h = (bar_plot_h * RAIN_TIER_TOP_PCT_ARR[tier]) / 100;
    return h > 0 ? h : 1;
}

int rain_tier_proportional_height(int tenths, int bar_plot_h) {
    if (tenths <= 0 || bar_plot_h <= 0) {
        return 0;
    }
    const int tier    = rain_tier_of_tenths(tenths);
    const int fill_q8 = rain_tier_fill_q8(tenths, tier);

    // Lower tiers contribute their full segment height; the topmost
    // tier contributes fill_q8/256 of its segment so the top edge moves
    // continuously across the wire-tenths domain.
    const int below_h          = (bar_plot_h * RAIN_TIER_TOP_PCT_ARR[tier - 1]) / 100;
    const int slab_top_full    = (bar_plot_h * RAIN_TIER_TOP_PCT_ARR[tier])     / 100;
    const int slab_h_full      = slab_top_full - below_h;
    int slab_h_top = (slab_h_full * fill_q8) / 256;
    if (slab_h_top == 0 && fill_q8 > 0) { slab_h_top = 1; }

    const int total = below_h + slab_h_top;
    return total > 0 ? total : 1;
}

const ChartColorStop *rain_tier_stops(int *num_stops) {
    return palette_rain_stops(num_stops);
}

GColor rain_tier_color(int tier) {
    int n = 0;
    const ChartColorStop *stops = palette_rain_stops(&n);
    if (tier <= 0 || n <= 0) { return PBL_IF_COLOR_ELSE(GColorWhite, GColorBlack); }
    int idx = tier - 1;
    if (idx >= n) { idx = n - 1; }   // B&W: single stop catches every tier
    return stops[idx].color;
}

int16_t rain_tier_permille(int tenths) {
    return (int16_t)rain_tier_proportional_height(tenths, 1000);
}

void rain_tier_fill_permille(const uint8_t *tenths, int16_t *out, int count) {
    for (int i = 0; i < count; ++i) {
        out[i] = rain_tier_permille(tenths[i]);
    }
}

File: src/pkjs/weather/rain-tier.js
// Authoritative rain-tier definition (single source of truth for colors).
// Height constants mirror C rain_tier.c (RAIN_TIER_MAX_TENTHS / RAIN_TIER_TOP_PCT_ARR);
// they are duplicated there for the radar's on-watch height math — keep in sync.
var COLORS = require('../pebble-colors');

var MAX_TENTHS = [1, 5, 20, 100];                 // tier upper bounds (wire tenths)
var TOP_PCT    = [0, 14, 34, 56, 78, 100];        // cumulative slab tops (% of plot)
// Per-tier colors for colour displays, tiers 1..5.
var TIER_COLORS = [
    COLORS.GColorLightGray, COLORS.GColorElectricBlue, COLORS.GColorGreen,
    COLORS.GColorYellow, COLORS.GColorSunsetOrange
];
// B&W platforms get a single black stop; the watch adds the white outline.
var BW_PLATFORMS = { aplite: true, diorite: true };

/**
 * Tier index 1..5 for a wire-tenths rain value, or 0 for <= 0.
 * @param {number} tenths Rain in wire tenths.
 * @returns {number} Tier index.
 */
function tierOfTenths(tenths) {
    if (tenths <= 0) { return 0; }
    for (var i = 0; i < MAX_TENTHS.length; i += 1) {
        if (tenths <= MAX_TENTHS[i]) { return i + 1; }
    }
    return 5;
}

/**
 * Fraction (0..256) of the topmost tier slab that is filled. Port of C
 * rain_tier_fill_q8.
 * @param {number} tenths Rain in wire tenths.
 * @param {number} tier Tier index 1..5.
 * @returns {number} q8 fill in [0,256].
 */
function fillQ8(tenths, tier) {
    var low, high;
    switch (tier) {
        case 1: return 256;
        case 2: low = 2;   high = 5;   break;
        case 3: low = 6;   high = 20;  break;
        case 4: low = 21;  high = 100; break;
        case 5: low = 101; high = 255; break;
        default: return 256;
    }
    if (tenths >= high) { return 256; }
    if (tenths <= low)  { return 0; }
    return Math.trunc(((tenths - low) * 256) / (high - low));
}

/**
 * Per-mille (0..1000) bar height for a wire-tenths rain value. Exact port of C
 * rain_tier_permille / rain_tier_proportional_height(tenths, 1000).
 * @param {number} tenths Rain in wire tenths.
 * @returns {number} Height in per-mille of plot height.
 */
function rainPermille(tenths) {
    if (tenths <= 0) { return 0; }
    var tier = tierOfTenths(tenths);
    var q8 = fillQ8(tenths, tier);
    var belowH = Math.trunc((1000 * TOP_PCT[tier - 1]) / 100);
    var slabTopFull = Math.trunc((1000 * TOP_PCT[tier]) / 100);
    var slabHFull = slabTopFull - belowH;
    var slabHTop = Math.trunc((slabHFull * q8) / 256);
    if (slabHTop === 0 && q8 > 0) { slabHTop = 1; }
    var total = belowH + slabHTop;
    return total > 0 ? total : 1;
}

/**
 * Build the rain color palette for a watch platform + bar-color choice.
 * B&W platforms always get a single black stop (the watch adds the white outline) and
 * ignore rainBarColor. On color displays, 'white' collapses to a single white stop;
 * anything else (default) yields the five multicolor tier stops.
 * @param {string} platform Pebble platform id (aplite/basalt/chalk/diorite/emery).
 * @param {string} [rainBarColor] 'multicolor' (default) or 'white'. Color displays only.
 * @returns {{from: number[], rgb: number[]}} Stops: permille thresholds + 0xRRGGBB colors.
 */
function buildPalette(platform, rainBarColor) {
    if (BW_PLATFORMS[platform]) {
        return { from: [0], rgb: [COLORS.GColorBlack] };
    }
    if (rainBarColor === 'white') {
        return { from: [0], rgb: [COLORS.GColorWhite] };
    }
    return {
        from: [0, 140, 340, 560, 780],   // TOP_PCT[0..4] * 10
        rgb: TIER_COLORS.slice()
    };
}

module.exports = {
    rainPermille: rainPermille,
    buildPalette: buildPalette
};
